### PR TITLE
[SPARK-43802][SQL] Fix codegen for unhex and unbase64 with failOnError=true

### PR DESCRIPTION
**Pull request by Kimahriman**
### What changes were proposed in this pull request?

Fixes an error with codegen for unhex and unbase64 expression when failOnError is enabled introduced in https://github.com/apache/spark/pull/37483.

### Why are the changes needed?

Codegen fails and Spark falls back to interpreted evaluation:
```
Caused by: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 47, Column 1: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 47, Column 1: Unknown variable or type "BASE64"
```
in the code block:
```
/* 107 */         if (!org.apache.spark.sql.catalyst.expressions.UnBase64.isValidBase64(project_value_1)) {
/* 108 */           throw QueryExecutionErrors.invalidInputInConversionError(
/* 109 */             ((org.apache.spark.sql.types.BinaryType$) references[1] /* to */),
/* 110 */             project_value_1,
/* 111 */             BASE64,
/* 112 */             "try_to_binary");
/* 113 */         }
```

### Does this PR introduce _any_ user-facing change?

Bug fix.

### How was this patch tested?

Added to the existing tests so evaluate an expression with failOnError enabled to test that path of the codegen.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala [hunk 1]
```diff
@@ -615,6 +615,9 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(Unhex(Literal("GG")), null)
     checkEvaluation(Unhex(Literal("123")), Array[Byte](1, 35))
     checkEvaluation(Unhex(Literal("12345")), Array[Byte](1, 35, 69))
+
+    // failOnError
+    checkEvaluation(Unhex(Literal("12345"), true), Array[Byte](1, 35, 69))
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:620]: Does this check that compilation doesn't fail, but not the error?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 2]
```diff
@@ -1172,14 +1172,13 @@ case class Unhex(child: Expression, failOnError: Boolean = false)
...
       val maybeFailOnErrorCode = if (failOnError) {
-        val format = UTF8String.fromString("BASE64");
         val binaryType = ctx.addReferenceObj("to", BinaryType, BinaryType.getClass.getName)
         s"""
            |if (${ev.value} == null) {
            |  throw QueryExecutionErrors.invalidInputInConversionError(
            |    $binaryType,
            |    $c,
-           |    $format,
+           |    UTF8String.fromString("HEX"),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1181]: uhh, the format value is not correct too. Could you add a test for the error, please. Like:
https://github.com/apache/spark/blob/45e5f2e375bec915e1683e6d2a222488ba831c91/sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala#L60-L71

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala (hunk 1 above)
Comment (Kimahriman) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:620]: Yes this is just a positive test case that evaluation still works with failOnError enabled. It looked like there were other tests (that you commented on below), checking the error occurs, but probably doesn't care that it falls back to eval during the process.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala (hunk 2 above)
Comment (Kimahriman) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1181]: Not sure what you mean. The format here was wrong in addition to the bad codegen, just fixed both. Assume this was a copy paste error from unbase64

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala:620]: > but probably doesn't care that it falls back to eval during the process.

Could you add a test w/o fallbacks to catch the incorrect format, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1181]: > Not sure what you mean. The format here was wrong in addition to the bad codegen, just fixed both.

I mean add a test which fails without your second fix: BASE64 -> HEX

Comment (Kimahriman) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1181]: This was being like constant folded so it was just being eval'd. Update the existing error test to do both codegen and interpreted, and made one for hex as well

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master/3.4.
Thank you, @Kimahriman.

**Comment by MaxGekk** [comment]
@Kimahriman The changes cause conflicts in branch-3.4. Could you backport your changes to 3.4, please.

Just in case, does branch-3.3 has the same issue? If so, please, backport the changes to 3.3 too.

**Comment by Kimahriman** [comment]
> Just in case, does branch-3.3 has the same issue? If so, please, backport the changes to 3.3 too.

No this was added in 3.4, though there was talks of potentially backporting the original PR with the issue https://github.com/apache/spark/pull/37483. I'll make the 3.4 backport PR.

**Pull request closed without merging**